keps/1136-provisioning-request-support/README.md

```diff
@@ -0,0 +1,185 @@
+
+Before admission, CA will check whether there is enough resources and
+provide them if their number is not sufficient (details
+depend on the exact engine used with `ProvisioningRequest)`.
+
+### Goals
+
+* Provide Kueue integration with `ProvisioningReqest`.
+* Define how users can configure what Kueue puts into `ProvisioningRequest`.
+* Provide a list of test cases to ensure that all corner cases are covered.
+
+### Non-Goals
+
+* Define how Cluster Autoscaler handles ProvisioningRequest.
+* Define underling cloud-specific behavior.
+
+## Proposal
+
+* Introduce a new controller in Kueue that will act as AdmissionCheck based on
+  the status of created `ProvisioningRequest`.
+
+* Introduce a new cluster-scoped CRD to configure how `ProvisioningRequest` should be used.
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to admit workloads only after ClusterAutoscaler running on my cloud provider
+expands a dedicated node group on which the workload will be run.
+
+#### Story 2
+
+I want to admit workloads only after a CheckCapacity request to ClusterAutoscaler
+succeeds.
+
+### Risks and Mitigations
+
+There doesn't seem to be much risks or mitigations. (Two phase admission process)[
+was added specifically for use cases like this.
+
+## Design Details
+
+The new ProvisioningRequest controller will:
+
+* Watch for all workloads that require an `AdmissionCheck` with controller
+name set to `"ProvisioningRequestController"`. For that it will also need to
+to watch all `AdmissionCheck` definitions to understand whether the particular
+check is in fact `ProvisioningRequest` or not.
+
+* For each of such workloads create a `ProvisioningRequest` (and accompanying
+PodTemplates) requesting capacity for all podsets from the workload. Annotate
+the workload with the ProvisioningRequest name under:
+`"cluster-autoscaler.kubernetes.io/consume-provisioning-request"`
+To understand what details should it put into `ProvisioningRequest` the controller
+will also need to watch `ProvisioningRequestConfigs`.
+
+* Watch all changes CA makes to `ProvisioningRequests`. If the `Provisioned`
+or `CapacityAvailable` condition is set to `True` then finish the `AdmissionCheck`
+with success (and propagate the information about `ProvisioningRequest` name to
+workload pods - KEP #1145). If the `ProvisioningRequest` fails, fail the `AdmissionCheck`.
+
+* Watch the admission of the workload - if it is again suspended or finished, 
+the provisioning request should also be deleted (the last one can be achieved via
+OwnerReference)
+
+The definition of `ProvisioningRequestConfig` is relatively simple and is based on
+what can be set in `ProvisioningRequest`.
+
+```
+type ProvisioningRequestConfig struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+    
+    ProvisioningClass string
+    Parameters map[string]string
+} 
+```
+
+`AdmissionCheck` will point to this configuration:
+
+```
+AdmissionCheck:
```

Comment:
```suggestion
kind: AdmissionCheck
```
?

Make sure this matches the AdmissionCheck definition to avoid confusion.

Reply:
Done.